File: fc-chat/demo/part3/ch4/message-system-add-push/src/main/java/net/prostars/messagesystem/handler/websocket/FetchUserInvitecodeRequestHandler.java
package net.prostars.messagesystem.handler.websocket;

import net.prostars.messagesystem.constant.IdKey;
import net.prostars.messagesystem.constant.MessageType;
import net.prostars.messagesystem.dto.domain.UserId;
import net.prostars.messagesystem.dto.websocket.inbound.FetchUserInvitecodeRequest;
import net.prostars.messagesystem.dto.websocket.outbound.ErrorResponse;
import net.prostars.messagesystem.dto.websocket.outbound.FetchUserInvitecodeResponse;
import net.prostars.messagesystem.service.ClientNotificationService;
import net.prostars.messagesystem.service.UserService;
import org.springframework.stereotype.Component;
import org.springframework.web.socket.WebSocketSession;

@Component
@SuppressWarnings("unused")
public class FetchUserInvitecodeRequestHandler
    implements BaseRequestHandler<FetchUserInvitecodeRequest> {

  private final UserService userService;
  private final ClientNotificationService clientNotificationService;

  public FetchUserInvitecodeRequestHandler(
      UserService userService, ClientNotificationService clientNotificationService) {
    this.userService = userService;
    this.clientNotificationService = clientNotificationService;
  }

  @Override
  public void handleRequest(WebSocketSession senderSession, FetchUserInvitecodeRequest request) {
    UserId senderUserId = (UserId) senderSession.getAttributes().get(IdKey.USER_ID.getValue());
    userService
        .getInviteCode(senderUserId)
        .ifPresentOrElse(
            inviteCode ->
                clientNotificationService.sendMessage(
                    senderSession, senderUserId, new FetchUserInvitecodeResponse(inviteCode)),
            () ->
                clientNotificationService.sendMessage(
                    senderSession,
                    senderUserId,
                    new ErrorResponse(
                        MessageType.FETCH_USER_INVITECODE_REQUEST,
                        "Fetch user invite code failed.")));
  }
}

File: fc-chat/demo/part3/with-push/src/main/kotlin/com/yscorp/withpush/messagesystem/handler/websocket/FetchUserInvitecodeRequestHandler.kt
package com.yscorp.withpush.messagesystem.handler.websocket

import com.yscorp.withpush.messagesystem.constant.IdKey
import com.yscorp.withpush.messagesystem.constant.MessageType
import com.yscorp.withpush.messagesystem.dto.domain.UserId
import com.yscorp.withpush.messagesystem.dto.websocket.inbound.FetchUserInvitecodeRequest
import com.yscorp.withpush.messagesystem.dto.websocket.outbound.ErrorResponse
import com.yscorp.withpush.messagesystem.dto.websocket.outbound.FetchUserInvitecodeResponse
import com.yscorp.withpush.messagesystem.service.ClientNotificationService
import com.yscorp.withpush.messagesystem.service.UserService
import org.springframework.stereotype.Component
import org.springframework.web.socket.WebSocketSession

@Component
@Suppress("unused")
class FetchUserInvitecodeRequestHandler(
    private val userService: UserService,
    private val clientNotificationService: ClientNotificationService
) : BaseRequestHandler<FetchUserInvitecodeRequest> {

    /**
     * 사용자의 초대코드를 조회하고 성공/실패에 따라 클라이언트에 메시지를 전송합니다.
     */
    override fun handleRequest(senderSession: WebSocketSession, request: FetchUserInvitecodeRequest) {
        val senderUserId = senderSession.attributes[IdKey.USER_ID.value] as? UserId ?: return

        val inviteCode = userService.getInviteCode(senderUserId)

        if (inviteCode != null) {
            clientNotificationService.sendMessage(
                senderSession,
                senderUserId,
                FetchUserInvitecodeResponse(inviteCode)
            )
        } else {
            clientNotificationService.sendMessage(
                senderSession,
                senderUserId,
                ErrorResponse(
                    MessageType.FETCH_USER_INVITECODE_REQUEST,
                    "Fetch user invite code failed."
                )
            )
        }
    }
}
